### PR TITLE
feat: update API/SFC preference using shortcuts

### PR DESCRIPTION
**Pull request by edimitchel**
## Description of Problem
Easy the switch without using mouse and keep focus on content

## Proposed Solution
Add keyup event listener for API and SFC preference.
firefox_ZIBbwDdyPO

>Ctrl+(Alt/Option)+A: toggle Options/Composition API
><s>Ctrl+(Alt/Option)+T: toggle template (SFC or HTML)</s>

## Additional Information
Externalize logic into a composable

### CODE REVIEW

**Comment by edimitchel** [comment]
@skirtles-code, this one!

**Comment by yyx990803** [comment]
Can we get rid of the dotted underline? It looks a bit out of place. We can mention the shortcut at the end of the "Introduction - API Styles" section .

**Comment by edimitchel** [comment]
> Can we get rid of the dotted underline? It looks a bit out of place. We can mention the shortcut at the end of the "Introduction - API Styles" section .

Yes, I'll do that

File: .vitepress/theme/components/preferences.ts [hunk 1]
```diff
@@ -19,3 +19,98 @@ export function filterHeadersByPreference(headers: Header[]) {
...
+  }
+
+  onMounted(() => {
+    document.addEventListener('keyup', preferenceKeyupHandler)
+  })
+  onUnmounted(() => {
+    document.removeEventListener('keyup', preferenceKeyupHandler)
+  })
+
+  const shortcutInfo = computed(() => {
```
Comment (dammy001) [inline .vitepress/theme/components/preferences.ts:101]: I think this computed value should be above the methods.

**Comment by edimitchel** [comment]
I just tried on OSX, and shortcut doesn't work for an unknown reason..

**Comment by edimitchel** [comment]
> I just tried on OSX, and shortcut doesn't work for an unknown reason..

Fixed!

**Comment by dammy001** [comment]
@edimitchel Great 👍

File: src/guide/introduction.md [hunk 2]
```diff
@@ -200,7 +200,14 @@ If you are new to Vue, here's our general recommendation:
 
   - Go with Composition API + Single-File Components if you plan to build full applications with Vue.
 
-You don't have to commit to only one style during the learning phase. The rest of the documentation will provide code samples in both styles where applicable, and you can toggle between them at any time using the **API Preference switches** at the top of the left sidebar.
+You don't have to commit to only one style during the learning phase. The rest of the documentation will provide code samples in both styles where applicable, and you can toggle between them at any time using the **API Preference switches** at the top of the left sidebar. 
+
+> Use following shortcuts to switch between API Preference:
+> 
+> * `Ctrl+Alt+A` switches the API Preference (options/composition)
```
Comment (yyx990803) [inline src/guide/introduction.md:207]: Small improvement here: we can use a dynamic text binding for the `Alt / Option` part and set it to the correct key after component mount using similar logic [here](https://github.com/vuejs/theme/commit/02522c09906d7f07d11f0cc3911459adbcd7fa73)

**Review by yyx990803 — CHANGES_REQUESTED** [review]

File: src/guide/introduction.md (hunk 2 above)
Comment (edimitchel) [reply src/guide/introduction.md]: navigator.platform is not available from navigator nor from globalThis/window (crash on vitepress build)

Comment (edimitchel) [reply src/guide/introduction.md]: Should be fixed now. I take the opportunity to separate things

**Comment by edimitchel** [comment]
@yyx990803 ready for review

**Comment by edimitchel** [comment]
I use keyCode as key doesn't result to `a` or `t` but Dead and a cross symbol.

**Comment by edimitchel** [comment]
Any news here @yyx990803 ?

**Comment by edimitchel** [comment]
@NataliaTepluhina can we bring this into Vue docs?

**Comment by NataliaTepluhina** [comment]
@edimitchel will take a look this week when I am back from the conference! @skirtles-code do you have any comments on this one?

**Comment by edimitchel** [comment]
Preview URL: https://deploy-preview-1449--vuejs.netlify.app/
Netlify url in the comment above is not good

**Review by skirtles-code — CHANGES_REQUESTED** [review]
Here are a few small suggestions from me based on an initial read through.

I'd also note that I'm on Ubuntu, which uses <kbd>Ctrl+Alt+T</kbd> to launch a terminal window, so I can't actually use that shortcut to toggle the API. But I'm not sure whether that's worth addressing for the relatively smaller number of people affected.

**Review by skirtles-code — CHANGES_REQUESTED** [review]
GitHub seems to be having a lot of problems today. For some reason this review comment didn't get included with my review, so here it is separately:

File: src/guide/introduction.md [hunk 3]
```diff
@@ -213,7 +213,17 @@ If you are new to Vue, here's our general recommendation:
...
+
+<script>
+  import usePlatform from '/@theme/composables/usePlatform.ts'
+  const { altKey } = usePlatform();
+</script>
+
+> Use following shortcuts to switch between API Preference:
+> 
+> * `Ctrl+{{altKey}}+A` switches the API Preference (Options/Composition)  
+> * `Ctrl+{{altKey}}+T` switches the templating preference (HTML/SFC)
```
[the remark this answers is not in the document]
Comment (edimitchel) [reply src/guide/introduction.md:226]: Yes, and I think this won't be a useful shortcut. I remove it and only keep the API preference shortcut

**Comment by edimitchel** [comment]
> I'd also note that I'm on Ubuntu, which uses Ctrl+Alt+T to launch a terminal window, so I can't actually use that shortcut to toggle the API. But I'm not sure whether that's worth addressing for the relatively smaller number of people affected.

Decided to remove the SFC Preference shortcut which don't making sense, unlink the API Preference shortcut

File: .vitepress/theme/components/PreferenceSwitch.vue [hunk 4]
```diff
@@ -1,64 +1,34 @@
...
-  isOpen = !isOpen
-}
+  preferSFC,
+} from '../composables/usePreferences'
+
+const {
+  isOpen,
+  shortcutInfo,
+  showPreference,
+  showSFC,
```
Comment (skirtles-code) [inline .vitepress/theme/components/PreferenceSwitch.vue:13]: It seems that `showSFC` is no longer being returned, so the second switch is currently missing on the Tutorial and Examples pages.

File: .vitepress/theme/composables/usePlatform.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+export default function usePlatform() {
+  const isMac = /(Mac OS X)/i.test(globalThis.navigator?.userAgent)
+  const altKey = isMac ? 'Option' : 'Alt';
```
Comment (skirtles-code) [inline .vitepress/theme/composables/usePlatform.ts:3]:
```suggestion
  const altKey = isMac ? 'Option' : 'Alt'
```

**Comment by skirtles-code** [comment]
@edimitchel I'd like to get this over the line. Could you take a look at my latest comments above? There are also merge conflicts with the current `main` branch. Thanks.

**Comment by edimitchel** [comment]
Closing this PR as it's not adding as much functionality as I thought

**Pull request closed without merging**